lib/ansible/modules/cloud/vmware/vmware_guest_network.py

```diff
@@ -0,0 +1,491 @@
+                            if 'name' in network and nic_device.deviceInfo.summary != network['name']:
+                                nic_device.deviceInfo.summary = network['name']
+                                self.change_detected = True
+                            if 'manual_mac' in network and nic_device.macAddress != network['manual_mac']:
+                                if vm_obj.runtime.powerState != vim.VirtualMachinePowerState.poweredOff:
+                                    self.module.fail_json(msg='Expected power state is poweredOff to reconfigure MAC address')
+                                nic_device.addressType = 'manual'
+                                nic_device.macAddress = network['manual_mac']
+                                self.change_detected = True
+                            if self.change_detected:
+                                self.config_spec.deviceChange.append(nic_spec)
+                        elif network['state'].lower() == 'absent':
+                            nic_spec.operation = vim.vm.device.VirtualDeviceSpec.Operation.remove
+                            nic_spec.device = nic_device
+                            self.change_detected = True
+                            self.config_spec.deviceChange.append(nic_spec)
+                else:
+                    self.module.fail_json(msg='Unable to find the specified network adapter: %s' % network)
+
+    def reconfigure_vm_network(self, vm_obj):
+        network_list = self.sanitize_network_params()
+        # gather network adapter facts only
+        if (self.params['gather_network_facts'] is not None and self.params['gather_network_facts']) or len(network_list) == 0:
+            results = {'changed': False, 'failed': False, 'network_data': self.get_network_facts(vm_obj)}
+        # do reconfigure then gather facts
+        else:
+            self.get_network_config_spec(vm_obj, network_list)
+            try:
+                task = vm_obj.ReconfigVM_Task(spec=self.config_spec)
+                wait_for_task(task)
+            except vim.fault.InvalidDeviceSpec as e:
+                self.module.fail_json(msg="Failed to configure network adapter on given virtual machine due to invalid"
+                                          " device spec : %s" % to_native(e.msg),
+                                      details="Please check ESXi server logs for more details.")
+            except vim.fault.RestrictedVersion as e:
+                self.module.fail_json(msg="Failed to reconfigure virtual machine due to"
+                                          " product versioning restrictions: %s" % to_native(e.msg))
+            if task.info.state == 'error':
+                results = {'changed': self.change_detected, 'failed': True, 'msg': task.info.error.msg}
+            else:
+                network_facts = self.get_network_facts(vm_obj)
+                results = {'changed': self.change_detected, 'failed': False, 'network_data': network_facts}
```

Comment:
mandatory msg is missing

Reply:
'msg' is mandatory?

Reply:
Yes, 'msg' key and value should always present on *_exit() calls

Reply:
I'm not sure about this, I checked "AnsibleModule" in this file https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/basic.py, "exit_json()" does not require "msg" key in result, but "fail_json()" requires "msg" key. Thanks.

Reply:
My fault, I agree with that.

Reply:
Thanks for the comment, I'm also not quite sure about this before check that, Thanks.